lib/kube/proxy/sess_test.go

```diff
@@ -73,99 +72,140 @@ func TestSessionEndError(t *testing.T) {
+			req, err := generateExecRequest(
+				generateExecRequestConfig{
+					addr:          testCtx.KubeProxyAddress(),
+					podName:       podName,
+					podNamespace:  podNamespace,
+					containerName: podContainerName,
+					cmd:           containerCommmandExecute, // placeholder for commands to execute in the dummy pod
+					options:       streamOpts,
+				},
+			)
+			require.NoError(t, err)
 
-	exec, err := remotecommand.NewSPDYExecutor(userRestConfig, http.MethodPost, req.URL())
-	require.NoError(t, err)
-	err = exec.StreamWithContext(testCtx.Context, streamOpts)
-	require.Error(t, err)
+			exec, err := remotecommand.NewSPDYExecutor(userRestConfig, http.MethodPost, req.URL())
+			require.NoError(t, err)
+			err = exec.StreamWithContext(testCtx.Context, streamOpts)
+			require.Error(t, err)
 
-	// check that the session is ended with an error in audit log.
-	require.EventuallyWithT(t, func(t *assert.CollectT) {
-		eventsResultMutex.Lock()
-		defer eventsResultMutex.Unlock()
-		hasSessionEndEvent := false
-		hasSessionExecEvent := false
-		for _, event := range eventsResult {
-			if event.GetType() == events.SessionEndEvent {
-				hasSessionEndEvent = true
-			}
-			if event.GetType() != events.ExecEvent {
-				continue
-			}
+			if tt.recordingErr == nil {
+				// check that the session is ended with an error in audit log.
+				require.EventuallyWithT(t, func(t *assert.CollectT) {
+					eventsResultMutex.Lock()
+					defer eventsResultMutex.Unlock()
+					hasSessionEndEvent := false
+					hasSessionExecEvent := false
+					for _, event := range eventsResult {
+						if event.GetType() == events.SessionEndEvent {
+							hasSessionEndEvent = true
+						}
+						if event.GetType() != events.ExecEvent {
+							continue
+						}
 
-			execEvent, ok := event.(*apievents.Exec)
-			assert.True(t, ok)
-			assert.Equal(t, events.ExecFailureCode, execEvent.GetCode())
-			assert.Equal(t, strconv.Itoa(errorCode), execEvent.ExitCode)
-			assert.Equal(t, errorMessage, execEvent.Error)
-			hasSessionExecEvent = true
-		}
-		assert.Truef(t, hasSessionEndEvent, "session end event not found in audit log")
-		assert.Truef(t, hasSessionExecEvent, "session exec event not found in audit log")
-	}, 10*time.Second, 1*time.Second)
+						execEvent, ok := event.(*apievents.Exec)
+						assert.True(t, ok)
+						assert.Equal(t, events.ExecFailureCode, execEvent.GetCode())
+						if tt.recordingErr == nil {
+							assert.Equal(t, strconv.Itoa(errorCode), execEvent.ExitCode)
+							assert.Equal(t, errorMessage, execEvent.Error)
+						} else {
+							assert.Empty(t, execEvent.ExitCode)
+							assert.Equal(t, tt.recordingErr.Error(), execEvent.Error)
+						}
+						hasSessionExecEvent = true
+					}
+					assert.Truef(t, hasSessionEndEvent, "session end event not found in audit log")
+					assert.Truef(t, hasSessionExecEvent, "session exec event not found in audit log")
+				}, 10*time.Second, 1*time.Second)
+			} else {
+				require.Never(t, func() bool {
```

Comment:
Do we need the periodic checking or is this equivalent to sleep(1s) -> check?

Could we do this in a way that doesn't take 1s?

Reply:
unfortunately no. They are async and we need to wait for collection. The test takes 5s without this